File: src/main/java/io/github/liyze09/arms/network/packet/clientbound/LoginSuccess.java
package io.github.liyze09.arms.network.packet.clientbound;

import io.github.liyze09.arms.network.Connection;
import io.github.liyze09.arms.network.packet.ClientBoundPacketEncoder;
import io.github.liyze09.arms.network.packet.Packet;
import org.jetbrains.annotations.NotNull;

import static io.github.liyze09.arms.network.PackUtils.*;

public class LoginSuccess implements ClientBoundPacketEncoder<LoginSuccess.LoginSuccessBody> {
    private static final LoginSuccess INSTANCE = new LoginSuccess();
    public static LoginSuccess getInstance() {
        return INSTANCE;
    }
    @Override
    public Packet encode(@NotNull LoginSuccessBody msg, @NotNull Connection connection) {
        var buffer = connection.ctx.alloc().buffer();
        // UUID
        buffer.writeLong(msg.uuid.a());
        buffer.writeLong(msg.uuid.b());
        // Username
        var name = msg.username.getBytes();
        writeVarInt(name.length, buffer);
        buffer.writeBytes(name);

        writeVarInt(0, buffer);
        if (checkProtocolVersion(connection.getProtocolVersion(), 766)) {
            buffer.writeByte((byte) 0);
        }

        return Packet.of(0x02, buffer);
    }
    public record LoginSuccessBody (
        Connection.UUID uuid,
        String username
    ){
    }
}

File: src/main/kotlin/io/github/liyze09/arms/network/packet/clientbound/LoginSuccess.kt
package io.github.liyze09.arms.network.packet.clientbound

import io.github.liyze09.arms.network.Connection
import io.github.liyze09.arms.network.PackUtils.writeMCBoolean
import io.github.liyze09.arms.network.PackUtils.writeVarInt
import io.github.liyze09.arms.network.packet.Packet
import net.minecraftarm.common.UUID

object LoginSuccess : io.github.liyze09.arms.network.packet.ClientBoundPacketEncoder<LoginSuccessBody> {
    override fun encode(msg: LoginSuccessBody, connection: Connection): Packet {
        val buffer = connection.ctx.alloc().buffer()
        // UUID
        buffer.writeLong(msg.uuid.a)
        buffer.writeLong(msg.uuid.b)
        // Username
        val name = msg.username.toByteArray()
        buffer.writeVarInt(name.size)
        buffer.writeBytes(name)

        buffer.writeVarInt(0)

        buffer.writeMCBoolean(true)

        return Packet.of(0x02, buffer)
    }

}

@JvmRecord
data class LoginSuccessBody(
    val uuid: UUID,
    val username: String
)
